src/main/redux/sagas/app.ts

```diff
@@ -50,4 +61,101 @@ export function* init() {
         },
     );
 
+    app.on("will-quit", () => {
+
+        debug("#####");
+        debug("will-quit");
+        debug("#####");
+    });
+
+}
+
+export function exit() {
+    return spawn(function*() {
+
+        const beforeQuitEventChannel = getBeforeQuitChannel();
+        const windowAllClosedEventChannel = getWindowAllClosedEventChannel();
+        const quitEventChannel = getQuitEventChannel();
+        let shouldExit: boolean = process.platform !== "darwin" || IS_DEV;
+
+        /*
+        // events order :
+        - before-quit
+        - window-all-closed
+        - quit
+        */
+
+        yield takeSpawnEveryChannel(
+            beforeQuitEventChannel,
+            (e: Electron.Event) => {
+
+                // track ctrl-q/command-q
+                shouldExit = true;
+
+                e.preventDefault();
+
+                const libraryWin = getLibraryWindowFromDi();
+                libraryWin.close();
+
+                debug("#####");
+                debug("#####");
+                debug("#####");
+
+                debug("before-quit");
+
+                debug("#####");
+                debug("#####");
+                debug("#####");
+            },
+        );
+
+        yield takeSpawnEveryChannel(
+            windowAllClosedEventChannel,
+            function*() {
+
+                debug("#####");
+                debug("#####");
+                debug("#####");
+
+                debug("window-all-closed");
+
+                debug("#####");
+                debug("#####");
+                debug("#####");
+
+                // clear session in r2-navigator
+                yield all([
+                    call(clearSessions),
+                    call(needToPersistState),
+                ]);
+
+                yield put(streamerActions.stopRequest.build());
+
+                yield race({
+                    a: take(streamerActions.stopSuccess.ID),
+                    b: take(streamerActions.stopError.ID),
+                });
+
+                if (shouldExit) {
+                    app.exit(0);
```

Comment:
As a rule of thumb, Electron apps should be cleanly shutdown using `app.quit()` not `app.exit()` (unless in CLI mode), because of subtle differences on Mac, Windows, Linux related to process lifecycle. What is the motivation for using `app.exit()` here?

Reply:
i'm manually handle the thorium end cycle. Make a call to app.quit twice launches 'before-quit' and 'will-quit' event.
When the user press `ctrl-q`, electron call app.quit(). This function dispatch a 'before-quit' event and try to close every browser windows and finally close the main process.
In thorium the ctrl-q shortcut should be like a press of the library window close button. To catch and stop the propagation of app.quit process, I prevent default the event in 'before-quit' event. In this case I fully controls the end-cycle when a developer make a call to app.quit or an user close the app.

The end-cycle workd like this:
- library 'close' event is trapped
    - closes every render windows
- app event 'window-all-closed' is trapped
    - calls the thorium closed process
    - if 'before-quit' event has been previously called or os=linux/window -> forces app exit
    - else in mac wait app event 'appActivate' to recreate a library window.

Reply:
i tried to answer, but tell me how you see thorium end-cycle ?

Reply:
Thank you for the clarification, I think your analysis covers all the use cases in Mac Windows and Linux.